Public/Src/Tools/Execution.Analyzer/Analyzers.Partner/ContentPlacementAnalyzer.cs

```diff
@@ -0,0 +1,532 @@
+                                    bxlArtifact.NumOutputPips += executions.Count;
+                                    foreach (var exec in executions)
+                                    {
+                                        bxlArtifact.OutputPips.Add(exec);
+                                    }
+                                }
+                            }
+                        }
+                        // and save
+                        WriteOutput(bxlArtifact);
+                    }
+#pragma warning disable ERP022
+                    catch
+                    {
+                        // I dont really want to log this one, but is not bad to count the errors
+                        ++errors;
+                    }
+#pragma warning restore ERP022
+                    ++processed;
+                    if (processed % printStep == 0)
+                    {
+                        Console.WriteLine($"{samples.Count - processed} artifacts remaining ({errors} errors...)");
+                    }
+                }
+            }
+            finally
+            {
+                m_outputWriter.WriteEndArray();
+                Console.WriteLine($"{processed} artifacts printed, done...");
+            }
+
+            return 0;
+        }
+
+
+        public Dictionary<FileArtifact, FileContentInfo> SampleAccordingToOverallDistribution(double sampleSize, Random random)
+        {
+            if (m_fileContentMap.Count <= m_sampleSize)
+            {
+                return m_fileContentMap;
+            }
+            else
+            {
+                var fileArtifacts = new Dictionary<FileArtifact, FileContentInfo>();
+                foreach (var extensionEntry in m_countFilesPerExtension)
+                {
+                    // get counts for each type of file
+                    double weightedCount = Math.Ceiling(extensionEntry.Value * sampleSize / m_fileContentMap.Count);
+                    // if we have less that the size, then take all
+                    if (extensionEntry.Value < weightedCount)
+                    {
+                        foreach (var a in m_artifactsPerExtension[extensionEntry.Key])
+                        {
+                            fileArtifacts[a] = m_fileContentMap[a];
+                            if (fileArtifacts.Count >= m_sampleSizeHardLimit)
+                            {
+                                return fileArtifacts;
+                            }
+                        }
+                    }
+                    else
+                    {
+                        // now, take a random set of files for each extension
+                        var randomSet = new HashSet<int>();
+                        while (randomSet.Count < weightedCount)
+                        {
+                            randomSet.Add(random.Next(extensionEntry.Value));

```

Comment:
Merely observing here:
Is there any assumption that there is a minimal size? Because technically the random function can be rounded constatly to the same index, and then only a single sample would be returned.